[RNMobile] Fix BottomSheet.SubSheet/TextInput Conflict

SiobhyB opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/33828

`gutenberg-mobile:` https://github.com/wordpress-mobile/gutenberg-mobile/pull/3784

## Description

As outlined in https://github.com/WordPress/gutenberg/issues/33828, the text editor for adding alt text to an image (accessed from the image block's settings) isn't currently working as expected. At the time of writing, the keyboard is immediately dismissed with each keystroke, making it extremely difficult, bordering on impossible, to add alt text to an image.

[BottomSheetTextControl](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/mobile/bottom-sheet-text-control) is the component behind the image block's alt text settings and is built on [the BottomSheet.SubSheet component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js). There were changes to `BottomSheet.SubSheet` in https://github.com/WordPress/gutenberg/pull/33240/commits/d650b4f1df6e1699c7c45819a8afd4a2ec39a08e that led to a conflict with the `TextInput` component in `BottomSheetTextControl`, which brought this issue to the surface.

With this PR, I have refactored the changes in https://github.com/WordPress/gutenberg/pull/33240/commits/d650b4f1df6e1699c7c45819a8afd4a2ec39a08e in a way that enables the alt text entry to work as expected.  

## How has this been tested?

### Test Case 1: Alt Text Settings

<details>
<summary>The following steps outline an approach to test that the original bug is fixed with this PR.</summary>

1. Navigate to a new post then add an image block or tap on an existing one.
2. Access the image block's settings by tapping the cog/gear icon. 
3. Tap **Alt Text** and confirm that a new sub sheet opens with a text editor.
4. Enter a few lines into the text editor and confirm that it no longer closes with each keystroke.
</details>

### Test Case 2: Help Panel (Production Only)

<details>
<summary>As part of our testing, we should also verify that the new changes don't bring any unexpected side effects to other places where the component's used, such as the Help panel.</summary>

* Run the demo app.
* Open the menu on iOS by clicking the overflow menu icon in the ActionBar or by shaking the device on Android.
* Tap **Help**. The main help topics bottom sheet should be displayed.
* Tap one of the help topics and verify that the new sheet displays as expected.

</details>

### Test Case 3: Full Screen (Code Changes Required)

<details>
<summary>Lastly, we'll need to double-check that the "isFullScreen" parameter works as expected following the refactor. This will require code changes.</summary>

* Find a component in the app where [the BottomSheet.SubSheet component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js) is being utilised, such as [the BottomSheetTextControl component](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/bottom-sheet-text-control/index.native.js).
* Add `isFullScreen={ true }` to the list of parameters being accepted by `BottomSheet.SubSheet`.
* Navigate to the area in the app where the edited component is being used and verify that it's rendered as a sheet that takes up the full screen.
</details>

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/2998162/128023488-cb4a35b3-34eb-4478-acf9-9e8835fc1ab3.mov

## Types of changes

Bug fix (non-breaking change which fixes an issue), with the following high-level rundown of the main code changes:

* The `setIsFullScreen( isFullScreen )` method was moved to [a standalone `useEffect` hook](https://github.com/WordPress/gutenberg/pull/33845/files#diff-d16d37d09ff59dce57c087e978ac5536503cb1642ec2b0ecdba9536f5c695b95R26), which fires only when [the values of either `showSheet` or `isFullScreen` change](https://github.com/WordPress/gutenberg/pull/33845/files#diff-d16d37d09ff59dce57c087e978ac5536503cb1642ec2b0ecdba9536f5c695b95R30).
* Moving the method to the standalone hook meant [the BottomSheetScreen component could be removed](https://github.com/WordPress/gutenberg/pull/33845/files#diff-d16d37d09ff59dce57c087e978ac5536503cb1642ec2b0ecdba9536f5c695b95L20). It appeared to be the addition of this component that was causing a conflict with the `TextInput` component.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Remove 'BottomSheetScreen' component

The 'BottomSheetScreen' component was introduced in the following commit in order to correct a render error related to the 'setIsFullScreen' method: https://github.com/WordPress/gutenberg/pull/33240/commits/d650b4f1df6e1699c7c45819a8afd4a2ec39a08e#diff-d16d37d09ff59dce57c087e978ac5536503cb1642ec2b0ecdba9536f5c695b95R20

The way the component's currently rendering is causing a conflict with the TextInput component, however.

With this commit, the component's removed, along with the 'setIsFullScreen' method (the method will be re-introduced in the following commits).

SiobhyB committed:
Refactor 'setIsFullScreen' method

With this commit, the 'setIsFullScreen' method is added back to the component. This time, it's added to its own separate 'useEffect' function.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js:
```diff
@@ -17,24 +21,23 @@ const BottomSheetSubSheet = ( {
 	showSheet,
 	isFullScreen,
 } ) => {
-	const BottomSheetScreen = ( { onMount } ) => {
-		useEffect( onMount, [] );
-		return children ?? null;
-	};
```
My understanding from the PR description and testing is that the creation of this component _within_ the render of its parent `BottomSheetSubSheet` is the root cause of the issue, correct? 

I am not suggesting we change this PR, but I wanted to share the below alternative that is closer to the original implementation, purely for discussion's sake. It would appear merely hoisting `BottomSheetScreen` out of `BottomSheetSubSheet` does fix the original keyboard/focus issue. 

<details><summary>Alternative Approach</summary>

```diff
diff --git a/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js b/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js
index 8654dc918e..2af2c0f5da 100644
--- a/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js
+++ b/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js
@@ -11,17 +11,17 @@ import { createSlotFill, BottomSheetConsumer } from '@wordpress/components';
 
 const { Fill, Slot } = createSlotFill( 'BottomSheetSubSheet' );
 
+const BottomSheetScreen = ( { children, onMount } ) => {
+	useEffect( onMount, [] );
+	return children;
+};
+
 const BottomSheetSubSheet = ( {
 	children,
 	navigationButton,
 	showSheet,
 	isFullScreen,
 } ) => {
-	const BottomSheetScreen = ( { onMount } ) => {
-		useEffect( onMount, [] );
-		return children ?? null;
-	};
-
 	return (
 		<>
 			{ showSheet && (
@@ -33,7 +33,9 @@ const BottomSheetSubSheet = ( {
 									onMount={ () =>
 										setIsFullScreen( isFullScreen )
 									}
-								/>
+								>
+									{ children }
+								</BottomSheetScreen>
 							) }
 						</BottomSheetConsumer>
 					</SafeAreaView>

```

</details>

**Edit:** Updated the diff to showcase the difference between the original implementation, instead of the difference between this PR.

SiobhyB replied:
Thanks for this, David! 🙇‍♀️ You're right on what I believe to be the root cause of this issue (i.e. the creation of the child component within the parent). I appreciate you sharing that alternative approach, as hoisting the component isn't something I'd have come up with by myself, so will add it to my knowledge bank going forward. 

Are there any positives or negatives to either approach? I'm definitely open to learning and updating the PR if there is anything steering you towards or away from either approach. If it's the case that they're fairly similar, with no major advantages/disadvantages to one another, I'll go ahead to merge the current PR.

dcalhoun replied:
> Are there any positives or negatives to either approach? I'm definitely open to learning and updating the PR if there is anything steering you towards or away from either approach. If it's the case that they're fairly similar, with no major advantages/disadvantages to one another, I'll go ahead to merge the current PR.

No, I wouldn't say there is a distinct positive or negative with either approach. It is merely differently structured code from my perspective. 

That said, after reviewing this again, I believe [this wrapping `BottomSheetConsumer`](https://github.com/WordPress/gutenberg/blob/e53d768735bc3848f2d27c535394f904654c693e/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js#L37) may be duplicative of [the new `useContext` usage](https://github.com/WordPress/gutenberg/blob/e53d768735bc3848f2d27c535394f904654c693e/packages/components/src/mobile/bottom-sheet/sub-sheet/index.native.js#L24) and unnecessary. I.e. we could likely remove `BottomSheetConsumer` altogether. 

```jsx
// React Context accessed via Consumer with function as a child 
<BottomSheetConsumer>
	{({ setIsFullScreen }) => {
		// ...
	}}
</BottomSheetConsumer>

// React Context accessed via useContext Hook
const { setIsFullScreen } = useContext(BottomSheetContext);
```

For an explanation as to why, both of the above are accessing the React Context object. The former is using [the `Context.Consumer` approach](https://reactjs.org/docs/context.html#contextconsumer). The latter is using [the React Hook approach](https://reactjs.org/docs/hooks-reference.html#usecontext). Either is fine. I believe this project generally attempts to rely upon React Hooks now-a-days.

Currently, this PR has a lingering `BottomSheetConsumer` that is largely going unused.

SiobhyB replied:
Ah, excellent, that's clear, thanks for the explanation! I've gone ahead to remove code surrounding that unnecessary `BottomSheetConsumer` in https://github.com/WordPress/gutenberg/pull/33845/commits/523239866d5b502672695af6daa30453b57d26ac and https://github.com/WordPress/gutenberg/pull/33845/commits/b75e766a615f38091b8a8dca55a792ab007f8562. I also tested and can confirm the alt text settings still work with the update.

dcalhoun reviewed (approved):
Great work on this! Thank you! I verified the original keyboard/focus issue no longer occurs on an iPhone SE and Samsung Galaxy S20. 

However, I did notice additional typing issues that appear to have been present from the very original implementation, even before https://github.com/WordPress/gutenberg/commit/d650b4f1df6e1699c7c45819a8afd4a2ec39a08e was applied. 

1. Typing quickly can cause the cursor to jump to an incorrect location, which results in incorrect text being entered into the field. 
2. (Android only) After moving the cursor to a different location, tapping the delete/backspace key caused the cursor to jump backwards 1 character.

<details><summary>Android Bug Recording</summary>

https://user-images.githubusercontent.com/438664/128039499-995a5d9f-d3f6-4b90-9043-9c15aa11efe9.mp4

</details>

Given that these issues appear to be separate, I think we should merge this fix as-is. We can open a separate issue and fix. WDYT?

dcalhoun commented:
I have created the following two issues for the bugs that I found while testing this. 

* https://github.com/WordPress/gutenberg/issues/33856
* https://github.com/WordPress/gutenberg/issues/33857

SiobhyB committed:
Remove 'BottomSheetConsumer'

'BottomSheetConsumer' is no longer necessary as we're now using 'useContext' to pull in 'setIsFullScreen'.

Further discussion here: https://github.com/WordPress/gutenberg/pull/33845#discussion_r681939450

SiobhyB committed:
Refactor reference to 'children'

As 'BottomSheetConsumer' was removed in https://github.com/WordPress/gutenberg/pull/33845/commits/523239866d5b502672695af6daa30453b57d26ac, 'children' no longer needs to be returned as part of a function. In fact, this results in an error.

With this commit, the function surrounding the reference to 'children' is removed.

SiobhyB commented:
Thank you for the review and surfacing those bugs, @dcalhoun 🙇‍♀️ 

I'll go ahead with the merge domino when the tests are complete, unless anything else comes up for me to address.
